packages/app/contexts/TokenListContext.tsx

```diff
@@ -144,42 +129,58 @@ export const TokenListProvider = ({ children }: PropsWithChildren) => {
   }, [address, callArray, chainId, tokenList]);
 
   const setupTokenList = useCallback(async () => {
-    let mergedTokenlistTokens = defaultTokenList;
     setIsLoading(true);
     async function getTokenListData(tokenlistURL: string) {
-      const res = await fetch(tokenlistURL);
+      const { signal } = abortControllerRef.current;
+      const res = await fetch(tokenlistURL, { signal });
       if (!res.ok) {
         throw new Error("Failed to fetch tokenlist data");
       }
       return res.json();
     }
 
-    Promise.allSettled(
+    await Promise.allSettled(
       fetchTokenlistUrls.map((list) => getTokenListData(list))
     ).then((results) => {
       results.forEach((result) => {
         if (result.status === "fulfilled") {
-          mergedTokenlistTokens = mergeTokenlists(
-            mergedTokenlistTokens,
-            result.value.tokens
+          const mergedTokenlist = [...defaultTokenList, ...result.value.tokens];
+
+          const tokenListOfCurrentChainId = mergedTokenlist.filter(
+            (token: TokenFromTokenlist) => token.chainId === chainId
           );
+
+          const tokenUniqueAddresses: string[] = [];
+          const tokenListNoDuplicates: TokenFromTokenlist[] = [];
+
+          tokenListOfCurrentChainId.forEach((token) => {
+            if (!tokenUniqueAddresses.includes(token.address)) {
+              tokenUniqueAddresses.push(token.address);
+              tokenListNoDuplicates.push(token);
+            }
+          });
```

Comment:
I see we are filtering and then running a forEach. Can we filter inside the forEach? This way we don't need more than 1 loop through the token list.

Reply:
Also, maybe something like this

```
const tokenUniqueAddressesSet = new Set<string>();
const tokenListNoDuplicates = tokenListOfCurrentChainId.filter((token) => {
  if (!tokenUniqueAddressesSet.has(token.address)) {
    tokenUniqueAddressesSet.add(token.address);
    return true;
  }
  return false;
});
```

and filter per chain too. This might be a cleaner solution